tests/MongoDB.Driver.Tests/PipelineDefinitionBuilderTests.cs

```diff
@@ -114,6 +114,72 @@ public void Merge_should_add_expected_stage()
             stages[0].Should().Be("{ $merge : { into : { db : 'database', coll : 'collection' } } }");
         }
 
+        [Fact]
+        public void UnionWith_should_add_expected_stage()
+        {
+            var pipeline = new EmptyPipelineDefinition<BsonDocument>();
+            var withCollection = Mock.Of<IMongoCollection<BsonDocument>>(
+                coll => coll.CollectionNamespace == CollectionNamespace.FromFullName("db.test"));
+
+            var unionWithPipeline = new EmptyPipelineDefinition<BsonDocument>()
+                .AppendStage<BsonDocument, BsonDocument, BsonDocument>("{ $match : { b : 1 } }");
+            var result = pipeline.UnionWith<BsonDocument, BsonDocument, BsonDocument>(
+                withCollection,
+                unionWithPipeline);
+
+            var stages = RenderStages(result, BsonDocumentSerializer.Instance);
+            stages[0].Should().Be("{ $unionWith : { coll : 'test', pipeline : [{ '$match' : { b : 1 } }] } }");
+        }
+
+        [Fact]
+        public void UnionWith_should_throw_when_pipeline_is_null()
+        {
+            PipelineDefinition<BsonDocument, BsonDocument> pipeline = null;
+            IMongoCollection<BsonDocument> withCollection = null;
+
+            var exception = Record.Exception(
+                () => pipeline.UnionWith(
+                    withCollection,
+                    new EmptyPipelineDefinition<BsonDocument>()
+            ));
+
+            var argumentNullException = exception.Should().BeOfType<ArgumentNullException>().Subject;
+            argumentNullException.ParamName.Should().Be("pipeline");
+        }
+
+        [Fact]
+        public void UnionWith_should_throw_when_TWith_is_not_the_same_with_TInput_and_pipeline_is_null()
+        {
+            var pipeline = new EmptyPipelineDefinition<BsonDocument>();
+            var collection = Mock.Of<IMongoCollection<BsonValue>>(
+                coll => coll.CollectionNamespace == CollectionNamespace.FromFullName("db.test"));
+
+            var result = pipeline.UnionWith<BsonDocument, BsonDocument, BsonValue>(
+                collection,
+                withPipeline: null);
+            var exception = Record.Exception(() => RenderStages(result, BsonDocumentSerializer.Instance));
```

Comment:
Seeing this test made me realize that this error is being checked too late. It should be caught when UnionWith is called and not later when Render is called.

This test should be:

    var pipeline = new EmptyPipelineDefinition<BsonDocument>();
    var withCollection = Mock.Of<IMongoCollection<object>>(
        coll => coll.CollectionNamespace == CollectionNamespace.FromFullName("db.test"));

    var exception = Record.Exception(() => pipeline.UnionWith(withCollection, withPipeline: null));

    var e = exception.Should().BeOfType<ArgumentException>().Subject;
    e.Message.Should().StartWith("The withPipeline cannot be null when TWith != TInput. A pipeline is required to transform the TWith documents to TInput documents.");
    e.ParamName.Should().Be("withPipeline");